src/components/Calendars/ClientScheduleAppointment.tsx

```diff
@@ -0,0 +1,432 @@
+'use client';
+
+import { Availability } from '@/types/Availability';
+import React, { useState, useEffect, useCallback } from 'react';
+
+type ClientScheduleAppointmentProps = {
+    availabilities: Availability[];
+};
+
+export default function ClientScheduleAppointment({ availabilities }: ClientScheduleAppointmentProps) {
+    const MONTH_NAMES = [
+        'Janeiro',
+        'Fevereiro',
+        'Março',
+        'Abril',
+        'Maio',
+        'Junho',
+        'Julho',
+        'Agosto',
+        'Setembro',
+        'Outubro',
+        'Novembro',
+        'Dezembro',
+    ];
+
+    const DAYS = ['Dom', 'Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb'];
+
+    const [month, setMonth] = useState(new Date().getMonth());
+    const [day, setDay] = useState(new Date().getDate());
+    const [year, setYear] = useState(new Date().getFullYear());
+    const [no_of_days, setNumDays] = useState<number[]>([]);
+    const [blankdays, setBlankDays] = useState<number[]>([]);
+    const [datepickerValue, setDatepickerValue] = useState('');
+    const [selectedHour, setSelectedHour] = useState<number | null>();
+
+    const [step, setStep] = useState(0);
+    const [availabilitiesMap, setAvailabilities] = useState<Map<string, number[]>>();
+    const [availableHours, setAvailableHours] = useState<number[]>();
+
+    // Calculates the number of days in the given month
+    const getNoOfDays = useCallback(
+        (month: number) => {
+            let daysInMonth = new Date(year, month + 1, 0).getDate();
+            let dayOfWeek = new Date(year, month).getDay();
+
+            let blankdaysArray: number[] = [];
+            for (var i = 1; i <= dayOfWeek; i++) {
+                blankdaysArray.push(i);
+            }
+
+            let daysArray: number[] = [];
+            for (var i = 1; i <= daysInMonth; i++) {
+                daysArray.push(i);
+            }
+
+            setBlankDays(blankdaysArray);
+            setNumDays(daysArray);
+        },
+        [year],
+    );
+
+    useEffect(() => {
+        getNoOfDays(month);
+    }, [month, getNoOfDays]);
+
+    useEffect(() => {
+        if (availabilities) {
+            const datesList = availabilities.map((d: Availability) => d.attributes.date);
+            const dateMap: Map<string, number[]> = new Map();
+
+            for (let d of datesList) {
+                const date = new Date(d).toLocaleDateString();
+                const hour = parseInt(d.substr(11, 2));
+                if (!dateMap.has(date)) {
+                    dateMap.set(date, []);
+                }
+                const hoursArray = dateMap.get(date);
+                if (hoursArray) hoursArray.push(hour);
+            }
+
+            setAvailabilities(dateMap);
+            setAvailableHours(dateMap.get(new Date().toLocaleDateString()));
+        }
+    }, [availabilities]);
```

Comment:
Again, this doesn't need to be an use effect because you are mapping over what you receive from the props.

This function is very confusing to me as well, why are you getting all the dates, and then mapping them to a map of <date_str, hour>. Could you not use the dates as is? Maintaining the date split like this can be troublesome eventually

Reply:
My thought was, if i keep a Map of <date_str, hour[]> it will be much easier to check what are the hours available for every date. Knowing the date  and having a map prevents me from iterating an array everytime  i need to get the available hours

Reply:
its a solid way of doing it, I suggest this way of just keeping date as the state as it will result into having less problems when you start dealing with dates, to always have them around :D